pennylane/pauli/conversion.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Utility functions to convert between ``~.PauliSentence`` and other PennyLane operators.
+"""
+from functools import reduce, singledispatch
+
+from pennylane.operation import Tensor
+from pennylane.ops import Hamiltonian, Identity, PauliX, PauliY, PauliZ, Prod, SProd, Sum
+
+from .pauli_arithmetic import PauliWord, PauliSentence, X, Y, Z, op_to_str_map
+from .utils import is_pauli_word
+
+
+@singledispatch
+def pauli_sentence(op):
+    """Return the PauliSentence representation of an arithmetic operator or Hamiltonian.
+
+    Args:
+        op (~.Operator): The operator or Hamiltonian that needs to be converted.
+
+    Raises:
+        ValueError: Op must be a linear combination of Pauli operators
+
+    Returns ~.PauliSentence
+    """
+    raise ValueError(f"Op must be a linear combination of Pauli operators only, got: {op}")
+
+
+@pauli_sentence.register
+def _(op: PauliX):
+    return PauliSentence({PauliWord({op.wires[0]: X}): 1.0})
+
+
+@pauli_sentence.register
+def _(op: PauliY):
+    return PauliSentence({PauliWord({op.wires[0]: Y}): 1.0})
+
+
+@pauli_sentence.register
+def _(op: PauliZ):
+    return PauliSentence({PauliWord({op.wires[0]: Z}): 1.0})
+
+
+@pauli_sentence.register
+def _(op: Identity):  # pylint:disable=unused-argument
+    return PauliSentence({PauliWord({}): 1.0})
+
+
+@pauli_sentence.register
+def _(op: Tensor):
+    if not is_pauli_word(op):
+        raise ValueError(f"Op must be a linear combination of Pauli operators only, got: {op}")
+
+    pw = {}
+    for factor in op.obs:
+        wire, pauli = (factor.wires[0], op_to_str_map[factor.__class__])
+        pw[wire] = pauli
+
+    return PauliSentence({PauliWord(pw): 1.0})
+
```

Comment:
The conversion of a `Tensor` to a `PauliSentence` here produces an odd result if it's being used in Tensors with overlapping wires, ie.

```
op = qml.PauliX(0) @ qml.PauliY(0)
pauli_sentence(op)
>>> 1.0 * Y(0)
```
Same thing for a `Hamiltonian` defined based on a `Tensor` of multiple Pauli operators acting on the same wire.

If we fix #3334 so that `is_pauli_word` confirms that an operator is truly a Pauli word, this will raise a ValueError instead (and we should reword the error message a little). That would be fine, but it would be inconsistent with what happens if you try to do the same thing with `Prod`, which handles it nicely:

```
op = prod(qml.PauliX(0) @ qml.PauliY(0))
pauli_sentence(op)
>>> 1j * Z(0)
```

Would it be weird to convert it into a `Prod` instead of processing it as a `Tensor`? `Prod` behaves better and the output is all `PauliWord` and `PauliSentence` objects either way.

Reply:
We definitely need to fix `is_pauli_word` but I think we should do that in a separate PR. For now I update the logic to take this case into account when constructing from a tensor and add test cases for this.

Reply:
Added 👍🏼